qi-doc/scribblings/principles.scrbl

```diff
@@ -55,6 +67,107 @@ Consider this example:
+
+That brings us to the guarantee that Qi provides in this case (and in general).
+
+@subsection{Qi's Guarantee on Effects}
+
+@definition["Effect locality"]{For @tech{flow} invocations @${f} and @${g} and corresponding effects @${ε(f)} and @${ε(g)},
+
+@$${f < g ⇒ ε(f) < ε(g)}
+
+where @${<} on the left denotes the relation of being upstream, and @${<} on the right denotes one effect happening before another.
+}
+
+This has a few implications of note.
+
+@subsubsection{The @racket[effect] Form}
+
+First, for the @racket[effect] form, this implies that Qi considers an effectful flow @${f} performing an effect @${e} to be indistinguishable from a flow @racket[(effect e f′)], where @${f′} is @${f} without @${e} (and therefore pure), and effects declared in this way will never be separated from the associated flows. Thus, Qi @seclink["Separate_Effects_from_Other_Computations"]{encourages writing pure functions} while preserving the intuitive association of effects with functions.
+
+@subsubsection{The @racket[esc] Form}
+
+Qi will not optimize a flow that is wrapped with @racket[esc]. Thus, such flows will exhibit @seclink["Racket_vs_Qi"]{Racket's order of effects} (which, as we'll discuss below, also satisfies the requirements of locality).
+
+For more on how @racket[esc] is handled, see @secref["Using_Racket_to_Define_Flows"].
+
+@subsubsection{Truncating Effects}
+
+Next, for a flow like this one:
+
+@racketblock[
+  (~>> (filter my-odd?) (map my-sqr) car)
+]
+
+… when it is invoked with a large input list, Qi in fact only processes the very first value of the list, since it determines, at the end, that no further elements are needed in order to generate the final result. This means that all effects on would-be subsequent invocations of @racket[my-odd?] and @racket[my-sqr] would simply not be performed. Yet, locality is preserved here, since the @techlink[#:key "effect locality"]{defining implication} holds for every flow invocation that actually happens. Locality is about effects being guided by the @emph{necessity} of @techlink[#:key "associated effect"]{associated} computations to the final result.
+
+@subsubsection{A Natural Order of Effects}
+
+By being as fine-grained as possible in expressing computational dependencies, and in tying the execution of effects to such computational dependencies, @tech{effect locality} is in some sense the minimum well-formed guarantee on effects, and a natural one for functional languages to provide.
+
+@subsection{Racket vs Qi}
+
+In the earlier example, with an input list @racket[(list 1 2 3)], Racket's order of effects follows the invocation order:
+
+@racketblock[
+  (my-odd? 1) (my-odd? 2) (my-odd? 3) (my-sqr 1) (my-sqr 3)
+]
+
+Qi's order of effects is:
+
+@racketblock[
+  (my-odd? 1) (my-sqr 1) (my-odd? 2) (my-odd? 3) (my-sqr 3)
+]
```

Comment:
It may be helpful to repeat the programs here for context.

Reply:
Yes, it helps a lot!